cirq/contrib/quirk/cells/parse.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import (
+    Any,
+    Union,
+)
+
+import numpy as np
+import sympy
+from sympy.parsing.sympy_parser import parse_expr
+
+
+def parse_formula(formula: Any) -> Union[float, sympy.Basic]:
+    if not isinstance(formula, str):
+        raise TypeError('formula must be a string: {!r}'.format(formula))
+
+    formula = expand_unicode_fractions(formula)
+    try:
+        result = parse_expr(formula)
+    except Exception as ex:
+        raise ValueError(
+            f'Failed to parse the gate formula {repr(formula)}.\n'
+            'This is likely due to differences in how sympy and Quirk parse.\n'
+            'For example, Quirk allows "2 pi" whereas sympy requires "2*pi"\n'
+            'Parsing of sympy-incompatible formulas is not supported yet.'
+        ) from ex
+    if not result.free_symbols <= {sympy.Symbol('t')}:
+        raise ValueError(
+            f'Formula has variables besides time "t": {repr(formula)}')
+    if not result.free_symbols:
+        result = float(result)
+    return result
+
+
+UNICODE_FRACTIONS = {
+    "½": "(1/2)",
+    "¼": "(1/4)",
+    "¾": "(3/4)",
+    "⅓": "(1/3)",
+    "⅔": "(2/3)",
+    "⅕": "(1/5)",
+    "⅖": "(2/5)",
+    "⅗": "(3/5)",
+    "⅘": "(4/5)",
+    "⅙": "(1/6)",
+    "⅚": "(5/6)",
+    "⅐": "(1/7)",
+    "⅛": "(1/8)",
+    "⅜": "(3/8)",
+    "⅝": "(5/8)",
+    "⅞": "(7/8)",
+    "⅑": "(1/9)",
+    "⅒": "(1/10)",
+}
+
+
+def expand_unicode_fractions(text: str) -> str:
+    for k, v in UNICODE_FRACTIONS.items():
+        text = text.replace('√' + k, f'(sqrt{v})')
+        text = text.replace(k, v)
+    return text
+
+
+def parse_matrix(text: str) -> np.ndarray:
+    if not text.startswith('{{') or not text.endswith('}}'):
+        raise ValueError('No opening/closing braces.\ntext: {!r}'.format(text))
+    text = expand_unicode_fractions(text[2:-2])
+    rows = text.split('},{')
+    return np.array([[_parse_complex(c) for c in row.split(',')] for row in rows
+                    ])
+
+
+def _parse_complex(text: str) -> complex:
+    try:
+        if (text.endswith('i') and len(text) > 1 and not text.endswith('+i') and
+                not text.endswith('-i')):
+            text = text[:-1] + '*i'
+        expr = sympy.parsing.sympy_parser.parse_expr(text)
+        return complex(expr.subs({'i': 1j}))
```

Comment:
You don't need to make the substitution. Sympy [supports](https://docs.sympy.org/latest/_modules/sympy/parsing/sympy_parser.html) imaginary unit using capital i. Something like this should work and also avoids the need for preprocessing by the conditional statement above (l.87-89):

```python
return complex(parse_expr(text.replace('+i', '+I').replace('-i', '-I').replace('i', '*I'))) 
```

(That said, I really think we should reconsider using such a versatile parser, so this is probably moot.)

Reply:
I've completely rewritten this method. It now passes the fact that 'i' maps to 1j in via a dictionary to the parse method. It also restricts the available functions and constants to the set available in cirq. It furthermore unifies both the complex and formula parsing through the same code.